File: scripts/update_instagram_avatar.mjs
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const USERNAME = "decusangelicum";
const PROFILE_URL = `https://www.instagram.com/${USERNAME}/`;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, "..");
const ASSETS_DIR = path.join(ROOT, "assets");
const AVATAR_FILE = path.join(ASSETS_DIR, "avatar.jpg");
const META_FILE = path.join(ASSETS_DIR, "avatar-meta.json");

const headers = {
  "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
  "accept-language": "en-US,en;q=0.9,es;q=0.8",
  "cache-control": "no-cache",
};

function extractOgImage(html) {
  const patterns = [
    /<meta\s+property="og:image"\s+content="([^"]+)"/i,
    /<meta\s+content="([^"]+)"\s+property="og:image"/i,
  ];
  for (const pattern of patterns) {
    const match = html.match(pattern);
    if (match?.[1]) return match[1].replaceAll("&amp;", "&");
  }
  return null;
}

async function readMaybe(file) {
  try {
    return await readFile(file);
  } catch {
    return null;
  }
}

async function readMeta() {
  const raw = await readMaybe(META_FILE);
  if (!raw) return {};
  try {
    return JSON.parse(raw.toString("utf-8"));
  } catch {
    return {};
  }
}

async function writeMeta(meta) {
  await writeFile(META_FILE, `${JSON.stringify(meta, null, 2)}\n`, "utf-8");
}

async function run() {
  await mkdir(ASSETS_DIR, { recursive: true });

  try {
    let imageUrl = null;
    let imageBuffer = null;
    let source = "instagram-og-image";

    const profileRes = await fetch(PROFILE_URL, { headers });
    if (profileRes.ok) {
      const html = await profileRes.text();
      imageUrl = extractOgImage(html);
    }

    if (imageUrl) {
      const imageRes = await fetch(imageUrl, {
        headers: { ...headers, referer: PROFILE_URL },
      });
      if (imageRes.ok) {
        imageBuffer = Buffer.from(await imageRes.arrayBuffer());
      }
    }

    if (!imageBuffer || imageBuffer.byteLength < 10_000) {
      throw new Error("Instagram bloqueó/limitó la descarga del avatar (sin fallback de logo)");
    }
    const oldBuffer = await readMaybe(AVATAR_FILE);
    const changed = !oldBuffer || !oldBuffer.equals(imageBuffer);

    if (changed) {
      await writeFile(AVATAR_FILE, imageBuffer);
    }

    const now = Math.floor(Date.now() / 1000);
    const meta = await readMeta();
    const next = {
      ...meta,
      username: USERNAME,
      source,
      sourceProfile: PROFILE_URL,
      sourceImageUrl: imageUrl,
      lastChecked: now,
      avatarVersion: changed ? now : (meta.avatarVersion ?? now),
    };
    delete next.lastError;
    await writeMeta(next);

    console.log(changed ? "Avatar actualizado." : "Sin cambios en avatar.");
    return 0;
  } catch (error) {
    const now = Math.floor(Date.now() / 1000);
    const meta = await readMeta();
    const next = {
      ...meta,
      username: USERNAME,
      source: "avatar-updater",
      sourceProfile: PROFILE_URL,
      lastChecked: now,
      lastError: String(error?.message ?? error),
    };
    await writeMeta(next);

    console.error(`Error actualizando avatar: ${next.lastError}`);
    return 1;
  }
}

const code = await run();
process.exit(code);

File: scripts/update_instagram_avatar.py
import json
import re
import time
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

USERNAME = "decusangelicum"
ROOT = Path(__file__).resolve().parents[1]
ASSETS = ROOT / "assets"
AVATAR_FILE = ASSETS / "avatar.jpg"
META_FILE = ASSETS / "avatar-meta.json"
PROFILE_URL = f"https://www.instagram.com/{USERNAME}/"


def fetch_text(url: str) -> str:
    req = Request(
        url,
        headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
            "Accept-Language": "en-US,en;q=0.9,es;q=0.8",
            "Cache-Control": "no-cache",
        },
    )
    with urlopen(req, timeout=20) as res:
        return res.read().decode("utf-8", errors="replace")


def fetch_bytes(url: str) -> bytes:
    req = Request(
        url,
        headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
            "Referer": PROFILE_URL,
            "Cache-Control": "no-cache",
        },
    )
    with urlopen(req, timeout=20) as res:
        return res.read()


def extract_og_image(html: str) -> str | None:
    patterns = [
        r'<meta property="og:image" content="([^"]+)"',
        r"<meta content=\"([^\"]+)\" property=\"og:image\"",
    ]
    for p in patterns:
        m = re.search(p, html)
        if m:
            return m.group(1).replace("&amp;", "&")
    return None


def load_meta() -> dict:
    if not META_FILE.exists():
        return {}
    try:
        return json.loads(META_FILE.read_text(encoding="utf-8"))
    except Exception:
        return {}


def save_meta(meta: dict) -> None:
    META_FILE.write_text(json.dumps(meta, indent=2, ensure_ascii=False), encoding="utf-8")


def main() -> int:
    ASSETS.mkdir(parents=True, exist_ok=True)

    try:
        html = fetch_text(PROFILE_URL)
        image_url = extract_og_image(html)
        if not image_url:
            raise RuntimeError("No se encontró og:image en el perfil.")

        image_bytes = fetch_bytes(image_url)
        if len(image_bytes) < 10_000:
            raise RuntimeError("Imagen descargada demasiado pequeña; posible respuesta inválida.")

        old_bytes = AVATAR_FILE.read_bytes() if AVATAR_FILE.exists() else b""
        changed = image_bytes != old_bytes

        if changed:
            AVATAR_FILE.write_bytes(image_bytes)

        meta = load_meta()
        meta.update(
            {
                "username": USERNAME,
                "source": "instagram-og-image",
                "sourceProfile": PROFILE_URL,
                "lastChecked": int(time.time()),
                "avatarVersion": int(time.time()) if changed else meta.get("avatarVersion", int(time.time())),
            }
        )
        save_meta(meta)

        print("Avatar actualizado." if changed else "Sin cambios en avatar.")
        return 0

    except (HTTPError, URLError, TimeoutError, RuntimeError) as e:
        print(f"Error actualizando avatar: {e}")
        # Keep existing avatar; only update heartbeat metadata
        meta = load_meta()
        meta.update(
            {
                "username": USERNAME,
                "source": "instagram-og-image",
                "sourceProfile": PROFILE_URL,
                "lastChecked": int(time.time()),
                "lastError": str(e),
            }
        )
        save_meta(meta)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
